compiler/lib/src/main/scala/codegen/CppWriter/TopologyCppWriter/TopComponentInstances.scala

```diff
@@ -10,43 +10,45 @@ case class TopComponentInstances(
   aNode: Ast.Annotated[AstNode[Ast.DefTopology]]
 ) extends TopologyCppWriterUtils(s, aNode) {
 
-  private val bannerComment = "Component instances"
+  def getMembers: List[CppDoc.Member] = {
+    val instanceMembers = getInstanceMembers
+    List.concat(
+      guardedList (!instanceMembers.isEmpty) (List(getCommentMember)),
+      instanceMembers
+    )
+  }
 
-  def getHppLines: List[Line] = addBannerComment(
-    bannerComment,
-    getDeclLines
-  )
+  private val bannerComment = "Component instances"
 
-  def getCppLines: List[Line] = addBannerComment(
-    bannerComment,
-    getDefLines
+  private def getCommentMember = linesMember(
+    CppDocWriter.writeBannerComment(bannerComment),
+    CppDoc.Lines.Both
   )
 
-  private def getDeclLines = {
-    def getCode(ci: ComponentInstance): List[Line] = {
-      val implType = getImplType(ci)
-      val instanceName = ci.getUnqualifiedName
-      val instLines = lines(
-        s"""|//! $instanceName
-            |extern $implType $instanceName;"""
-      )
-      wrapInNamespaceLines(ci.qualifiedName.qualifier, instLines)
-    }
-    flattenWithBlankPrefix(instances.map(getCode))
-  }
-
-  private def getDefLines = {
-    def getCode(ci: ComponentInstance): List[Line] = {
+  private def getInstanceMembers = {
+    def getMembers(ci: ComponentInstance): List[CppDoc.Member] = {
       val implType = getImplType(ci)
       val instanceName = ci.getUnqualifiedName
-      val instLines = getCodeLinesForPhase (CppWriter.Phases.instances) (ci).getOrElse(
+      val hppMember = linesMember(
         lines(
-          s"$implType $instanceName(FW_OPTIONAL_NAME($q$instanceName$q));"
-        )
+          s"""|
+              |//! $instanceName
+              |extern $implType $instanceName;"""
+        ),
+        CppDoc.Lines.Hpp
       )
-      wrapInNamespaceLines(ci.qualifiedName.qualifier, instLines)
+      val cppMember = {
+        val instLines = getCodeLinesForPhase (CppWriter.Phases.instances) (ci).getOrElse(
+          lines(
+            s"""|
+                |$implType $instanceName(FW_OPTIONAL_NAME($q$instanceName$q));"""
+          )
+        )
+        linesMember(instLines, CppDoc.Lines.Cpp)
+      }
+      wrapInNamespaces(ci.qualifiedName.qualifier, List(hppMember, cppMember))
```

Comment:
will this re-introduce the possibility of empty namespaces?

Reply:
No, because there is one namespace member with an hpp member and a cpp member. The hpp member gets written to the hpp file, and the cpp member gets written to the cpp file. This is how namespace members are designed to work.

Before we had two namespace members, one with an hpp member and one with a cpp member. In this case the first namespace member had nothing to write to the cpp file, and the second one had nothing to write to the hpp file.

Reply:
In pseudocode, the CppDoc data structure is something like this:
```
namespace member with name "M" {
  lines member with lines "extern U32 x;" to be written to hpp
  lines member with lines "U32 x;" to be written to cpp
}
```
It generates the following code in the hpp file:
```
namespace M {
  extern U32 x;
}
```
And in the cpp file:
```
namespace M {
  U32 x;
}
```

Reply:
Before we had
```
namespace member with name "M" {
  lines member with lines "extern U32 x;" to be written to hpp
}
namespace member with name "M" {
  lines member with lines "U32 x;" to be written to cpp
}
```
That generates
```
namespace M {
  extern U32 x;
}
namespace M {

}
```
in hpp and
```
namespace M {

}
namespace M {
  U32 x;
}
```
in cpp.

Reply:
I see, thanks!